experimental/ActionPolyRing/src/Content.jl

```diff
@@ -0,0 +1,1373 @@
+
+function __update_internals!(dpre::DifferencePolyRingElem)
+  upr = __upr(parent(dpre))
+  p = data(dpre)
+  if upr.mpoly_ring !== parent(data(p))
+    dpre.p = upr(p)
+  end
+end  
+
+function __update_internals!(dpre::DifferentialPolyRingElem)
+  upr = __upr(parent(dpre))
+  p = data(dpre)
+  if upr.mpoly_ring !== parent(data(p))
+    dpre.p = upr(p)
+  end
+end  
+
+###############################################################################
+#
+#  Aux action polynomial rings 
+#
+###############################################################################
+
+### Difference ###
+
+#Getters for internals
+data(dpre::DifferencePolyRingElem) = dpre.p
+
+__upr(dpr::DifferencePolyRing) = dpr.upoly_ring
+
+__jtv(dpr::DifferencePolyRing) = dpr.jet_to_var
+
+__vtj(dpr::DifferencePolyRing) = dpr.var_to_jet
+
+__jtu_idx(dpr::DifferencePolyRing) = dpr.jet_to_upoly_idx
+
+__are_perms_up_to_date(dpr::DifferencePolyRing) = dpr.are_perms_up_to_date
+
+function __perm_for_sort(dpr::DifferencePolyRing)
+  if __are_perms_up_to_date(dpr) && isdefined(dpr, :permutation)
+    return dpr.permutation
+  end
+  __set_perm_for_sort!(dpr)
+  return dpr.permutation
+end
+
+function __perm_for_sort_poly(dpre::DifferencePolyRingElem)
+  dpr = parent(dpre)
+  if __are_perms_up_to_date(dpr) && isdefined(dpre, :permutation)
+    return dpre.permutation
+  end
+  __set_perm_for_sort!(dpr)
+  __set_perm_for_sort_poly!(dpre)
+  return dpre.permutation
+end
+
+#Setters for internals
+function __set_are_perms_up_to_date!(dpr::DifferencePolyRing, update::Bool)
+  dpr.are_perms_up_to_date = update
+end
+
+function __set_perm_for_sort!(dpr::DifferencePolyRing)
+  dpr.permutation = sortperm(dpr.(gens(__upr(dpr))); rev = true)
+  __set_are_perms_up_to_date!(dpr, true)
+end
+
+#If this is called, we assume that are_perms_up_to_date == true
+function __set_perm_for_sort_poly!(dpre::DifferencePolyRingElem)
+  __update_internals!(dpre)
+  dpre.permutation = sortperm(map(expo -> expo[parent(dpre).permutation], collect(exponents(data(dpre)))); rev = true)
+end
+
+### Differential ###
+data(dpre::DifferentialPolyRingElem) = dpre.p
+
+__upr(dpr::DifferentialPolyRing) = dpr.upoly_ring
+
+__jtv(dpr::DifferentialPolyRing) = dpr.jet_to_var
+
+__vtj(dpr::DifferentialPolyRing) = dpr.var_to_jet
+
+__jtu_idx(dpr::DifferentialPolyRing) = dpr.jet_to_upoly_idx
+
+__are_perms_up_to_date(dpr::DifferentialPolyRing) = dpr.are_perms_up_to_date
+
+function __perm_for_sort(dpr::DifferentialPolyRing)
+  if __are_perms_up_to_date(dpr) && isdefined(dpr, :permutation)
+    return dpr.permutation
+  end
+  __set_perm_for_sort!(dpr)
+  return dpr.permutation
+end
+
+function __perm_for_sort_poly(dpre::DifferentialPolyRingElem)
+  dpr = parent(dpre)
+  if __are_perms_up_to_date(dpr) && isdefined(dpre, :permutation)
+    return dpre.permutation
+  end
+  __set_perm_for_sort!(dpr)
+  __set_perm_for_sort_poly!(dpre)
+  return dpre.permutation
+end
```

Comment:
I assume that all of the above is the same for Difference* and Differential*. If that is indeed the case, please use `Union{DifferencePolyRing, DifferentialPolyRing}` in the signatures to reduce code duplication and to make it clear that both implementations are identical

For future readers of this code, some comments would probably be helpful